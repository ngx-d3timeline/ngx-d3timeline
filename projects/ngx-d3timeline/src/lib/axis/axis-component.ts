import { Component, Input } from '@angular/core';
import { Axis } from './axis';

@Component({
  selector: '[ngx-d3timeline-axis]',
  template: `
    <svg:g class="axis-group">
      <svg:g
        *ngFor="let tickMark of axis.tickMarks"
        ngx-d3timeline-axis-tick-mark
        [tickMark]="tickMark"
      ></svg:g>
      <svg:g ngx-d3timeline-line [line]="axis.line"></svg:g>
    </svg:g>
  `
})
export class AxisComponent {
  @Input() axis: Axis;
}
