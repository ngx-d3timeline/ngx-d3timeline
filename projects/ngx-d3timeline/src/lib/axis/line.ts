import { Orientation } from '../orientation';

export interface Line {
  x1: number;
  x2: number;
  y1: number;
  y2: number;
}

export function createLine(
  x1: number,
  y1: number,
  length: number,
  orientation: Orientation
) {
  return orientation === Orientation.Vertical
    ? createVerticalLine(x1, y1, length)
    : createHorizontalLine(x1, y1, length);
}

export function createVerticalLine(x1: number, y1: number, length: number) {
  return {
    x1,
    y1,
    x2: x1,
    y2: y1 + length
  };
}

export function createHorizontalLine(x1: number, y1: number, length: number) {
  return {
    x1,
    y1,
    x2: x1 + length,
    y2: y1
  };
}
