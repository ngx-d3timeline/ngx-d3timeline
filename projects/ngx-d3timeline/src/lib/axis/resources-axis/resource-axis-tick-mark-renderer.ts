import { BandScale } from '../../scale-types';
import { TickMarkRenderer, tickLabelSpacing } from '../tick-mark-renderer';
import { Orientation } from '../../orientation';

export function getResourceAxisTickMarkRenderer(
  scale: BandScale,
  orientation: Orientation
): TickMarkRenderer {
  const tickLineOffset = 0;

  function getBandMidPoint(tickValue: string): number {
    return scale(tickValue) + scale.bandwidth() / 2;
  }

  return {
    getTickLabel: (tickValue: string) => tickValue,
    getTickLabelSpacing: () => tickLineOffset + tickLabelSpacing,
    getTickValues: () => scale.domain(),
    mapTickValueToPositionInScale: (tickValue: string) =>
      getBandMidPoint(tickValue),
    tickLineOffset,
    orientation
  };
}
