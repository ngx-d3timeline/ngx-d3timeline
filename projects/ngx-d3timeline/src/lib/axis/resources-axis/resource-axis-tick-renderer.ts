import { BandScale } from '../../scale-types';
import { TickRenderer } from '../tick-renderer';

export class ResourceAxisTickRenderer implements TickRenderer<BandScale> {
  getTickValues(scale: BandScale): string[] {
    return scale.domain();
  }

  getLabel(_: BandScale, tick: any): string {
    return tick;
  }

  getTransform(scale: BandScale, tick: any): number {
    return this.getBandMidPoint(scale, tick);
  }

  private getBandMidPoint(scale: BandScale, tick: string): number {
    return scale(tick) + scale.bandwidth() / 2;
  }

  getTickLineOffset(): number {
    return 0;
  }
}
