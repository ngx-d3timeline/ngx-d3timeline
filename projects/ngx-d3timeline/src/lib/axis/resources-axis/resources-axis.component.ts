import { Component, ChangeDetectionStrategy } from '@angular/core';
import { AxisService } from '../axis.service';

@Component({
  selector: '[ngx-d3timeline-resources-axis]',
  template: `
    <svg:g
      class="resources-axis-group"
      *ngIf="axisService.resourceAxis$ | async as axis"
    >
      <text
        *ngFor="let tick of axis.ticks"
        [attr.transform]="tick.transform"
        dy="-2"
      >
        {{ tick.label }}
      </text>
      <svg:g ngx-d3timeline-line [line]="axis.axisLine"></svg:g>
    </svg:g>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ResourcesAxisComponent {
  constructor(public axisService: AxisService) {}
}
