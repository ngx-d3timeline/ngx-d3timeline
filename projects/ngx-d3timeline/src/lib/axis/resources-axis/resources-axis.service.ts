import { Axis } from '../axis';
import { Orientation } from '../../orientation';
import { map } from 'rxjs/operators';
import { ScaleBand } from 'd3-scale';
import { Injectable } from '@angular/core';
import { Tick } from '../tick';
import { ScalesService } from '../../scales.service';
import { OptionsService } from '../../options.service';
import { AxisService } from '../axis.service';

@Injectable({ providedIn: 'root' })
export class ResourcesAxisService {
  vm$ = this.scalesService.scales$.pipe(
    map(scales =>
      this.createAxis(scales.scaleBand, scales.state.timelineOrientation)
    )
  );

  constructor(
    private scalesService: ScalesService,
    private optionsService: OptionsService,
    private axisService: AxisService
  ) {}

  private createAxis(
    scaleBand: ScaleBand<string>,
    timelineOrientation: Orientation
  ): Axis {
    const orientation = this.optionsService.flipOrientation(
      timelineOrientation
    );

    return {
      ticks: this.getTicks(scaleBand, orientation),
      axisLine: this.axisService.getAxisLine(scaleBand, orientation)
    };
  }

  private getTicks(scale: ScaleBand<string>, orientation: Orientation): Tick[] {
    return scale.domain().map(value => ({
      label: value,
      transform: this.optionsService.getTranslation(
        this.getBandMidPoint(scale, value),
        orientation
      )
    }));
  }

  private getBandMidPoint(scale: ScaleBand<string>, tick: string) {
    return scale(tick) + scale.bandwidth() / 2;
  }
}
