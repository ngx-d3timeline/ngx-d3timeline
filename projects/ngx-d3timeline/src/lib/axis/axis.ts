import { TickMark } from '../tick-mark/tick-mark';
import { Line } from './line';

export interface Axis {
  tickMarks: TickMark[];
  line: Line;
}
