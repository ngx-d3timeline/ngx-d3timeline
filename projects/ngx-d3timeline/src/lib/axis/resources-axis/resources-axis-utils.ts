import { TimelineView } from '../../view/timeline-view';
import { Orientation } from '../../orientation';
import { BandScale } from '../../scale-types';
import { ResourceAxisTickRenderer } from './resource-axis-tick-renderer';
import { createSelector } from '../../selector/memoized-selector';
import { selectView, selectResourceOrientation } from '../../store/state';
import { selectBandScale } from '../../store/timeline-selectors';
import { createAxis } from '../axis-utils';

function createResourceAxis(
  view: TimelineView,
  orientation: Orientation,
  bandScale: BandScale
) {
  return createAxis(
    view,
    orientation,
    bandScale,
    new ResourceAxisTickRenderer()
  );
}

export const selectResourceAxis = createSelector(
  [selectView, selectResourceOrientation, selectBandScale],
  createResourceAxis
);
