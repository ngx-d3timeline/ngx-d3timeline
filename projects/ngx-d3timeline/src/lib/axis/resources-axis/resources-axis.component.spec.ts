import { TestBed, ComponentFixture } from '@angular/core/testing';
import { ResourcesAxisComponent } from './resources-axis.component';
import { AxisService } from '../axis.service';
import { of } from 'rxjs';
import { Input, Component } from '@angular/core';
import { Line } from 'dist/ngx-d3timeline/lib/axis/line';
import { createLine } from '../line';
import { origin } from '../../point';

@Component({
  selector: '[ngx-d3timeline-line]',
  template: `
    <svg:g></svg:g>
  `
})
class FakeLineComponent {
  @Input() line: Line;
}

describe('ResourcesAxisComponent', () => {
  let fixture: ComponentFixture<ResourcesAxisComponent>;
  let axisService: AxisService;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [ResourcesAxisComponent, FakeLineComponent],
      providers: [{ provide: AxisService, useValue: { vm$: jest.fn() } }]
    });

    fixture = TestBed.createComponent(ResourcesAxisComponent);
    axisService = TestBed.inject(AxisService);
  });

  it('should not render if view model is null', () => {
    axisService.resourceAxis$ = of(null);
    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should render correctly', () => {
    const line = createLine(origin, { x: 10, y: 10 });

    axisService.resourceAxis$ = of({
      ticks: [
        {
          label: 'tick 1',
          transform: 'translate(0, 10)',
          line
        },
        {
          label: 'tick 2',
          transform: 'translate(0, 20)',
          line
        }
      ],
      line: { x1: 0, x2: 10, y1: 1, y2: 0 }
    });

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
