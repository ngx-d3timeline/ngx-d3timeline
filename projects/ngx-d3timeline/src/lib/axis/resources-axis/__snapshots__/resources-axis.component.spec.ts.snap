// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourcesAxisComponent should not render if view model is null 1`] = `
<div
  id="root0"
  ngx-d3timeline-resources-axis=""
>
  
</div>
`;

exports[`ResourcesAxisComponent should render correctly 1`] = `
<div
  id="root1"
  ngx-d3timeline-resources-axis=""
>
  
  <g
    class="resources-axis-group"
  >
    
    <text
      dy="-2"
      transform="translate(0, 10)"
    >
       tick 1 
    </text>
    <text
      dy="-2"
      transform="translate(0, 20)"
    >
       tick 2 
    </text>
    <g
      ngx-d3timeline-axis-line=""
    >
      <g />
    </g>
  </g>
</div>
`;
